dapp/src/node/mod.rs

```diff
@@ -0,0 +1,378 @@
+    AddExtensionLayer, Router, Server,
+};
+use ethers::{
+    core::k256::ecdsa::SigningKey,
+    prelude::{
+        transaction::eip2718::TypedTransaction, Address, Block, NameOrAddress, Signer, Transaction,
+        TransactionReceipt, TxHash, Wallet, H256, U256, U64,
+    },
+    utils::keccak256,
+};
+use evm_adapters::Evm;
+
+mod methods;
+use methods::{BoxedError, EthRequest, EthResponse};
+
+mod types;
+use types::{Error, JsonRpcRequest, JsonRpcResponse, ResponseContent};
+
+/// Configurations of the EVM node
+pub struct NodeConfig {
+    /// Chain ID of the EVM chain
+    chain_id: U64,
+    /// Signer accounts that will be initialised with `genesis_balance` in the genesis block
+    genesis_accounts: Vec<Wallet<SigningKey>>,
+    /// Native token balance of every genesis account in the genesis block
+    genesis_balance: U256,
+    /// Signer accounts that can sign messages/transactions from the EVM node
+    accounts: HashMap<Address, Wallet<SigningKey>>,
+    /// Configured block time for the EVM chain. Use `None` to mine a new block for every tx
+    automine: Option<Duration>,
+}
+
+impl Default for NodeConfig {
+    fn default() -> Self {
+        Self {
+            chain_id: U64::one(),
+            genesis_accounts: Vec::new(),
+            genesis_balance: U256::zero(),
+            accounts: HashMap::new(),
+            automine: None,
+        }
+    }
+}
+
+impl NodeConfig {
+    /// Returns the default node configuration
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Sets the chain ID
+    pub fn chain_id<U: Into<U64>>(mut self, chain_id: U) -> Self {
+        self.chain_id = chain_id.into();
+        self
+    }
+
+    /// Sets the genesis accounts
+    pub fn genesis_accounts(mut self, accounts: Vec<Wallet<SigningKey>>) -> Self {
+        self.genesis_accounts = accounts;
+        self
+    }
+
+    /// Sets the balance of the genesis accounts in the genesis block
+    pub fn genesis_balance<U: Into<U256>>(mut self, balance: U) -> Self {
+        self.genesis_balance = balance.into();
+        self
+    }
+
+    /// Sets the block time to automine blocks
+    pub fn automine<D: Into<Duration>>(mut self, block_time: D) -> Self {
+        self.automine = Some(block_time.into());
+        self
+    }
+}
+
+/// Node represents an EVM-compatible node designed for development environments. It serves an
+/// Ethereum-compatible JSON-RPC 2.0 server
+pub struct Node<S, E: Evm<S>> {
+    /// The node can wrap around a generic implementation of EVM
+    evm: E,
+    /// The node's configuration
+    config: NodeConfig,
+    /// The blockchain state of the node
+    blockchain: Blockchain,
+
+    _s: PhantomData<S>,
+}
+
+#[derive(Default)]
+/// Stores the blockchain data (blocks, transactions)
+pub struct Blockchain {
+    /// Mapping from block hash to the block number
+    blocks_by_hash: HashMap<H256, U64>,
+    /// Mapping from block number to the block
+    blocks: HashMap<U64, Block<TxHash>>,
```

Comment:
Should we make this a Vec? Also, should the Block be `Block<Transaction>` instead of `TxHash`?

Reply:
I was thinking, we can allow both `get_block` and `get_block_with_transactions`. In the latter case, we populate the transactions by fetching them from the `txs` map.